### PR TITLE
Adding DocTest to TrOCR

### PR DESCRIPTION
**Pull request by arnaudstiegler**
# Adding TrOCR to DocTests

For this model, there was actually a single docstring in the entire file (for the forward method).

A couple of comments:
- As far as I'm aware, there is no TF version of this model
- TrOCR is an edge case because it's meant to be used as the decoder for a VisionEncoderDecoder. So the forward function of the TrOCR is not meant to be called directly. As a result, I gave some example code to run a forward pass with TrOCR within a VisionEncoderDecoder 

Let me know if the docstring is relevant to the problem. I can also revert adapt it to actually just showcase the forward for the `TrOCRForCausalLM` outside of a VisionEncoderDecoder

@patrickvonplaten @ydshieh @patil-suraj

### CODE REVIEW

**Comment by ydshieh** [comment]
I took a quick look for now, and think it is indeed a nice addition, considering there is currently no example in `modeling_trocr.py` at all.
Thank you, @arnaudstiegler!

File: src/transformers/models/trocr/modeling_trocr.py [hunk 1]
```diff
@@ -891,13 +891,40 @@ def forward(
         Example:
 
         ```python
+        >>> from transformers import TrOCRProcessor, VisionEncoderDecoderModel
```
Comment (ydshieh) [inline src/transformers/models/trocr/modeling_trocr.py:894]: Would be great to avoid import `VisionEncoderDecoderModel` twice

File: src/transformers/models/trocr/modeling_trocr.py [hunk 2]
```diff
@@ -891,13 +891,40 @@ def forward(
...
         >>> from transformers import VisionEncoderDecoderModel, TrOCRForCausalLM, ViTModel, TrOCRConfig, ViTConfig
+        >>> import requests
+        >>> from PIL import Image
 
+        >>> processor = TrOCRProcessor.from_pretrained("microsoft/trocr-base-handwritten")
```
Comment (ydshieh) [inline src/transformers/models/trocr/modeling_trocr.py:899]: This is better to be after showing the dummy model, i.e. after the line

```
>>> model = VisionEncoderDecoderModel(encoder=encoder, decoder=decoder)
```

File: src/transformers/models/trocr/modeling_trocr.py [hunk 3]
```diff
@@ -891,13 +891,40 @@ def forward(
...
+
+        >>> # TrOCR is a decoder model and should be used within a VisionEncoderDecoderModel
```
Comment (ydshieh) [inline src/transformers/models/trocr/modeling_trocr.py:901]: Maybe add a comment here

```
>>> # init vision2text model with random weights
```

File: src/transformers/models/trocr/modeling_trocr.py [hunk 4]
```diff
@@ -891,13 +891,40 @@ def forward(
...
-
         >>> model = VisionEncoderDecoderModel(encoder=encoder, decoder=decoder)
+
+        >>> # If you want to start from the pretrained model, load it from a VisionEncoderDecoderModel
+        >>> model = VisionEncoderDecoderModel.from_pretrained("microsoft/trocr-base-handwritten")
+
+        >>> # load image from the IAM dataset
+        >>> url = "https://fki.tic.heia-fr.ch/static/img/a01-122-02.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw).convert("RGB")
+        >>> pixel_values = processor(image, return_tensors="pt").pixel_values
```
Comment (ydshieh) [inline src/transformers/models/trocr/modeling_trocr.py:912]: Maybe put this line just before 

```
>>> text = "hello world"
```

File: utils/documentation_tests.txt [hunk 5]
```diff
@@ -38,6 +38,7 @@ src/transformers/models/speech_encoder_decoder/modeling_speech_encoder_decoder.p
 src/transformers/models/speech_to_text/modeling_speech_to_text.py
 src/transformers/models/speech_to_text_2/modeling_speech_to_text_2.py
 src/transformers/models/swin/modeling_swin.py
+src/transformers/models/trocr/modeling_trocr.py
```
Comment (ydshieh) [inline utils/documentation_tests.txt:41]: nice ❤️!

Are you able to run the doctest locally and get it pass?

**Review by ydshieh — APPROVED** [review]
@arnaudstiegler

Very nice & clean! I left a few tiny comments, but this PR is ready to be merged!

File: src/transformers/models/trocr/modeling_trocr.py (hunk 1 above)
Comment (arnaudstiegler) [reply src/transformers/models/trocr/modeling_trocr.py]: Good catch! Removed it

File: src/transformers/models/trocr/modeling_trocr.py (hunk 2 above)
Comment (arnaudstiegler) [reply src/transformers/models/trocr/modeling_trocr.py]: Done

File: src/transformers/models/trocr/modeling_trocr.py (hunk 3 above)
Comment (arnaudstiegler) [reply src/transformers/models/trocr/modeling_trocr.py:898]: Done

File: src/transformers/models/trocr/modeling_trocr.py (hunk 4 above)
Comment (arnaudstiegler) [reply src/transformers/models/trocr/modeling_trocr.py:912]: Done

File: utils/documentation_tests.txt (hunk 5 above)
Comment (arnaudstiegler) [reply utils/documentation_tests.txt:41]: Yep, runs fine locally! I ran the test again after addressing your comments

**Comment by arnaudstiegler** [comment]
Addressed the comments and re-ran the test locally!
There's one thing though: `make fixup` messes up the formatting of the docstring, and removes the additional blank at the end of the docstring ([here](https://github.com/huggingface/transformers/blob/ae63ca7033aa1d6220a855498a5403d846ba51b3/src/transformers/models/trocr/modeling_trocr.py#L933-L934)). The issue is that without the additional line, doctest will fail when comparing outputs. Not sure why to be honest, but do you know how to prevent that `make fixup` from doing this? Otherwise, next time the file is touched, make mixup will introduce a bug in the docstring

**Comment by ydshieh** [comment]
Those blank lines issues should be treated by the places regarding `utils/prepare_for_doc_test.py` as shown in this guide [doc](https://github.com/huggingface/transformers/tree/main/docs#for-python-files).

This means you don't need to add extra blank line in the file. Did you run `python utils/prepare_for_doc_test.py` as in the guide before and after running the doctest?

**Comment by arnaudstiegler** [comment]
Got it! I did run the script and it seems like it's not being applied to the trocr (but I can see the added lines on other model files). I'll debug that today

**Comment by arnaudstiegler** [comment]
> OK, thank you. Don't hesitate to report it if you find this is some bug in `prepare_for_doc_test.py`.

Figured out the issue: there was some formatting issue in one of the docstring (see last commit) that prevented the script to correctly add a line. From `utils.prepare_for_doc_test.process_doc_file`
```
splits = code.split("```")
splits = [s if i % 2 == 0 else process_code_block(s, add_new_line=add_new_line) for i, s in enumerate(splits)]
```
splits were incorrect because of that, and the script wouldn't add an additional \n
Now test run locally, and make fixup has been run as well.
Should be good to go

**Comment by arnaudstiegler** [comment]
@ydshieh Is there anything else to do? I don't have writing access here, so I can't merge this

File: src/transformers/models/trocr/modeling_trocr.py [hunk 6]
```diff
@@ -891,13 +891,46 @@ def forward(
...
-        >>> from transformers import VisionEncoderDecoderModel, TrOCRForCausalLM, ViTModel, TrOCRConfig, ViTConfig
-
+        >>> from transformers import (
+        ...     TrOCRProcessor,
+        ...     VisionEncoderDecoderModel,
+        ...     TrOCRForCausalLM,
+        ...     ViTModel,
+        ...     TrOCRConfig,
+        ...     ViTConfig,
+        ... )
```
Comment (ydshieh) [inline src/transformers/models/trocr/modeling_trocr.py:901]: just a tiny thing, but would be really great if this could be in some order,  like

```python
        ...     TrOCRConfig,
        ...     TrOCRProcessor,
        ...     TrOCRForCausalLM,
        ...     ViTConfig,
        ...     ViTModel,
        ...     VisionEncoderDecoderModel,
```

File: src/transformers/models/trocr/modeling_trocr.py [hunk 7]
```diff
@@ -891,13 +891,46 @@ def forward(
...
+
+        >>> # TrOCR is a decoder model and should be used within a VisionEncoderDecoderModel
+        >>> # init vision2text model with random weights
         >>> encoder = ViTModel(ViTConfig())
         >>> decoder = TrOCRForCausalLM(TrOCRConfig())
-        # init vision2text model
-
         >>> model = VisionEncoderDecoderModel(encoder=encoder, decoder=decoder)
+
+        >>> # If you want to start from the pretrained model, load it from a VisionEncoderDecoderModel
```
Comment (ydshieh) [inline src/transformers/models/trocr/modeling_trocr.py:911]: Think would be better described as

```
>>> # If you want to start from the pretrained model, load the checkpoint with `VisionEncoderDecoderModel`
```

File: src/transformers/models/trocr/modeling_trocr.py [hunk 8]
```diff
@@ -891,13 +891,46 @@ def forward(
...
+        >>> # load image from the IAM dataset
+        >>> url = "https://fki.tic.heia-fr.ch/static/img/a01-122-02.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw).convert("RGB")
+        >>> text = "hello world"
+        >>> pixel_values = processor(image, return_tensors="pt").pixel_values
+
+        >>> # training
+        >>> model.config.decoder_start_token_id = processor.tokenizer.cls_token_id
+        >>> model.config.pad_token_id = processor.tokenizer.pad_token_id
+        >>> model.config.vocab_size = model.config.decoder.vocab_size
```
Comment (ydshieh) [inline src/transformers/models/trocr/modeling_trocr.py:924]:
```suggestion
        >>> model.config.vocab_size = model.config.decoder.vocab_size

```

Comment (ydshieh) [reply src/transformers/models/trocr/modeling_trocr.py:924]: (just a new line for style reason)

**Review by ydshieh — COMMENTED** [review]
Hi, @arnaudstiegler,

I have left a few more comments: most of them are just some nits.
This PR is really ready, and I will merge once you can apply the suggestions 💯 
Thank you so much!

File: src/transformers/models/trocr/modeling_trocr.py [hunk 9]
```diff
@@ -597,8 +597,8 @@ def forward(
 
                 If `past_key_values` are used, the user can optionally input only the last `decoder_input_ids` (those
                 that don't have their past key value states given to this model) of shape `(batch_size, 1)` instead of
-                all ``decoder_input_ids``` of shape `(batch_size, sequence_length)`. inputs_embeds (`torch.FloatTensor`
-                of shape `(batch_size, sequence_length, hidden_size)`, *optional*): Optionally, instead of passing
+                all `decoder_input_ids` of shape `(batch_size, sequence_length)`. inputs_embeds (`torch.FloatTensor` of
```
Comment (ydshieh) [inline src/transformers/models/trocr/modeling_trocr.py:600]: very nice!

File: src/transformers/models/trocr/modeling_trocr.py (hunk 7 above)
Comment (arnaudstiegler) [reply src/transformers/models/trocr/modeling_trocr.py]: Done

File: src/transformers/models/trocr/modeling_trocr.py (hunk 6 above)
Comment (arnaudstiegler) [reply src/transformers/models/trocr/modeling_trocr.py:901]: Done

File: src/transformers/models/trocr/modeling_trocr.py [hunk 10]
```diff
@@ -891,13 +891,46 @@ def forward(
...
+        >>> text = "hello world"
+        >>> pixel_values = processor(image, return_tensors="pt").pixel_values
+
+        >>> # training
+        >>> model.config.decoder_start_token_id = processor.tokenizer.cls_token_id
+        >>> model.config.pad_token_id = processor.tokenizer.pad_token_id
+        >>> model.config.vocab_size = model.config.decoder.vocab_size
+        >>> labels = processor.tokenizer(text, return_tensors="pt").input_ids
+        >>> outputs = model(pixel_values, labels=labels)
+        >>> loss = outputs.loss
```
[the remark this answers is not in the document]
Comment (arnaudstiegler) [reply src/transformers/models/trocr/modeling_trocr.py:931]: Added the loss. I actually changed the text to reflect the actual text on the image because the loss was super high with the dummy text (loss=22.00 vs loss=4.00). Let me know what's best

**Comment by arnaudstiegler** [comment]
Rebased on latest main and addressed the comments, the failing test is coming from main (failing on main as well)

**Comment by ydshieh** [comment]
Hi, the failed check seems irrelevant to your PR. No need to fix it here 😃. I will give a final look and merge. Thank you 💗

**Comment by arnaudstiegler** [comment]
Alright, thank you!

**Comment by ydshieh** [comment]
By the way, may I wonder why you choose to display 3 decimal for the loss value? O remember ee use 2 decimal in doc.py. I will check once I am available.

File: src/transformers/models/trocr/modeling_trocr.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/models/trocr/modeling_trocr.py:931]: Very nice for the change of target text! Awesome 👍

File: src/transformers/models/trocr/modeling_trocr.py [hunk 11]
```diff
@@ -891,13 +891,52 @@ def forward(
...
+        >>> # training
+        >>> model.config.decoder_start_token_id = processor.tokenizer.cls_token_id
+        >>> model.config.pad_token_id = processor.tokenizer.pad_token_id
+        >>> model.config.vocab_size = model.config.decoder.vocab_size
+
+        >>> labels = processor.tokenizer(text, return_tensors="pt").input_ids
+        >>> outputs = model(pixel_values, labels=labels)
+        >>> loss = outputs.loss
+        >>> round(loss.item(), 3)
+        5.303
```
Comment (ydshieh) [inline src/transformers/models/trocr/modeling_trocr.py:933]:
```suggestion
        >>> round(loss.item(), 2)
        5.30
```

File: src/transformers/models/trocr/modeling_trocr.py [hunk 12]
```diff
@@ -891,13 +891,52 @@ def forward(
...
+        ...     ViTConfig,
+        ...     ViTModel,
+        ...     VisionEncoderDecoderModel,
+        ... )
+        >>> import requests
+        >>> import torch
+        >>> from PIL import Image
+
+        >>> torch.manual_seed(0)  # doctest: +IGNORE_RESULT
+
```
Comment (ydshieh) [inline src/transformers/models/trocr/modeling_trocr.py:907]:
```suggestion

```

File: src/transformers/models/trocr/modeling_trocr.py [hunk 13]
```diff
@@ -891,13 +891,52 @@ def forward(
...
+        >>> from transformers import (
+        ...     TrOCRConfig,
+        ...     TrOCRProcessor,
+        ...     TrOCRForCausalLM,
+        ...     ViTConfig,
+        ...     ViTModel,
+        ...     VisionEncoderDecoderModel,
+        ... )
+        >>> import requests
+        >>> import torch
```
Comment (ydshieh) [inline src/transformers/models/trocr/modeling_trocr.py:903]:
```suggestion
```

**Review by ydshieh — COMMENTED** [review]
Just added 3 final suggestions

- No (real) need to set the seed
- display only 2 decimal numbers

This is more about to be aligned with the code sample in `doc.py`.

**Comment by arnaudstiegler** [comment]
Applied your changes, tested them locally, and ran make fixup as this was needed after the changes

**Comment by arnaudstiegler** [comment]
Ok, I tried rebasing on latest master and it doesn't seem to be doing the trick. Not sure what's causing that unfortunately, but unlikely to be due to the code changes I've done :)

**Comment by ydshieh** [comment]
Thank you again @arnaudstiegler ! (also for your patience)

I merge this PR now ❤️.
